Add overview, how it works, and glossary docs (#15)
divviup/public-docs: docs/glossary.md

[task](#task). Typically, app developers will integrate a DAP client library to
achieve this.

### Collector

An application that implements the collector subprotocol of
[DAP](#distributed-aggregation-protocol-dap). It collects aggregated batches
from the aggregators and provides the final aggregate to the
[subscriber](#subscriber). This is most often operated by the application
developer.

A library suitable for building collectors is found in the Rust crate
[`janus_collector`][janus_collector]. A primitive collector implementation is
available in [Janus][collect.rs].

### Distributed Aggregation Protocol (DAP)

A protocol that describes a multi-party distributed system for privacy
preserving aggregation of measurements. It is currently in the draft phase. Its
specification can be viewed on the [IETF data tracker][DAP].

Each participant in a DAP system adopts one of these roles: [client](#client),
[collector](#collector), [leader](#leader), or [helper](#helper).

### Helper

A [DAP](#distributed-aggregation-protocol-dap) aggregator that performs
aggregation and collection in response to direction from the Leader.

> The system requirements for a helper are lower than those for a leader.

Reviewer: nit: say "system and network bandwidth requirements" to match corresponding text in "Leader" section

### Leader

A [DAP](#distributed-aggregation-protocol-dap) aggregator that directly ingests
[reports](#report) from clients, and drives the aggregation and collection
processes with the helper. The leader determines how reports are to be batched.

The system and network bandwidth requirements for a leader are higher than for
those for a helper.

### Query type

Describes how [reports](#report) are grouped into batches for aggregation. There
are two query types defined by [DAP](#distributed-aggregation-protocol-dap):

- Time-interval: Reports are grouped into batches based on the timestamp of the
  report.

> - Fixed-size: Reports are grouped into arbitrary batches by the leader.

Reviewer: This should explain that the leader constructs batches of some fixed size. Arbitrary makes it sound capricious and unpredictable.

### Report

A client-submitted measurement. Each report is composed of two encrypted report

> shares: one for the leader, and one for the helper.

Author: More or less yes, as IIUC proof shares are specific to Prio3?

Perhaps more detail is justified here, especially since this is missing discussion of the public share. Let me take a stab at it.

> shares: one for the leader, and one for the helper.

Author: I added detail about the composition of the report, according to the struct definition https://www.ietf.org/archive/id/draft-ietf-ppm-dap-04.html#name-leader-initialization.

I incorporated discussion of proofs in step 4 of How It Works.

### Subscriber

An organization or individual who is using Divvi Up to implement privacy
perserving metrics for their application.

### Task

> Represents a single metric to be measured. The task is the core entity that

Author: I reworded this rather heavily, let me know how it looks.

---
# Glossary position should be as last as possible, but before the Resources category.
sidebar_position: 999
---

# Glossary

> Use the glossary to find definitions and further information for new or
> ambiguous terms.

Reviewer 2: ```suggestion
This glossary provides definitions of terms specific to Divvi Up and the protocols it uses.
```

### Aggregator

A server that implements the Distributed Aggregation Protocol (DAP). Divvi Up
and its partners operate DAP-compliant aggregators. Aggregators can be
self-hosted.

Aggregators ingest measurements sent from clients. They perform aggregation of

> batches or reports into aggregations, and service [collector](#collector)

Reviewer 2: I think this was a typo
```suggestion
batches of reports into aggregations, and service [collector](#collector)
```

requests.

Aggregators can be in one of two roles, [Leader](#leader) or [Helper](#helper).

Examples of aggregator software include [Janus (ISRG)][janus], and [Daphne
(Cloudflare, helper-only)][daphne].

### Client

> An end-user application whose developer wishes to collect metrics on.

Reviewer 2: I took a crack at rewriting this, but I don't know if I'm 100% happy with it.
```suggestion
Software that uses the DAP protocol to contribute a measurement to a metric. Typically, application developers will include a DAP client library in end-user software or a webapp.
```

Author: I landed on
> An end-user app that uses DAP to contribute a measurement to a metric. Typically, app developers can achieve this using a DAP client library.

### Collector

> An application that collects aggregate shares from aggregators and combines them

Author: I tweaked the sentence to `It collects aggregated batches from the aggregators and provides the final aggregate to the subscriber.`, but perhaps the wording could be better. I wanted to emphasize the finality of this, i.e. this is what a subscriber uses to get their final metric.

into the final aggregate. This is most often operated by the application
developer.

> A library suitable for building collectors is found in the crate

Reviewer 2: ```suggestion
A library suitable for building collectors is found in the Rust crate
```

[`janus_collector`][janus_collector]. A primitive collector implementation is
available in [Janus][collect.rs].

### Helper

A DAP aggregator that performs aggregation and collection in response to

> direction from the Leader. The system requirements for a helper are lower than a
> leader.

Reviewer 2: nit: parallelism
```suggestion
direction from the Leader. The system requirements for a helper are lower than
those for a leader.
```

### Leader

A DAP aggregator that directly ingests metrics, creates aggregation and
collection jobs, and drives those jobs to completion. The system and network

> bandwidth requirements for a leader are higher than for a helper.

Reviewer 2: ```suggestion
bandwidth requirements for a leader are higher than those for a helper.
```

### Query type

Describes how a batch of reports are to be aggregated. There are two query types
defined by DAP:

- Time-interval: Batches are queried based on some time range.

> - Fixed-size: Batches contain a fixed number of reports. The collector queries

Author: Given that we have a max_batch_size and min_batch_size, it's not strictly true that each batch would contain the same number of reports--this might mislead someone implementing a collector.

  for arbitrary batches.

### Report

> A client-submitted measurement.

Reviewer 2: Adding a little more detail:
```suggestion
A client-submitted measurement. Each report is composed of two report shares, one for the leader, and one for the helper.
```

### Subscriber

An organization or individual who is using Divvi Up to implement privacy
perserving metrics for their application.

### Task

> Represents the measurement process. The task the basic entity that Divvi Up and

Reviewer 3: ```suggestion
Represents the measurement process. The task is the core entity that Divvi Up and
```

> Represents the measurement process. The task is the core entity that Divvi Up and

Author: I suppose it would have been interesting if we called a "task" a "metric", but I'm not sure if that's overloading the term "metric" too much.
